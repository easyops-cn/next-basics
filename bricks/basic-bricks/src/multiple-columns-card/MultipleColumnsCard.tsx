import React from "react";
import { Card } from "antd";

export interface MultipleColumnsCardProps {
  cardBorder: boolean;
  gridColumns: string | Array<string | number>;
}

export function transformCss(
  gridColumns: MultipleColumnsCardProps["gridColumns"]
) {
  const defaultValue = "1fr";
  const arr = Array.isArray(gridColumns) ? gridColumns : [gridColumns];

  const columns = arr.reduce((str, value) => {
    if (typeof value === "string") {
      str += `${value} `;
    } else if (typeof value === "number") {
      str += `${value}fr `;
    }
    return str;
  }, "");

  return columns || defaultValue;
}

export function MultipleColumnsCard({
  cardBorder,
  gridColumns
}: MultipleColumnsCardProps): React.ReactElement {
  return (
    <Card bordered={cardBorder}>
      <div
        className="multiple-columns-container"
        style={{
          gridTemplateColumns: transformCss(gridColumns)
        }}
      >
        <slot id="contentSlot" name="content" />
      </div>
    </Card>
  );
}
