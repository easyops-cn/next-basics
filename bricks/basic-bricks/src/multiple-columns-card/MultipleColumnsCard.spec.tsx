import React from "react";
import { shallow } from "enzyme";
import { MultipleColumnsCard, transformCss } from "./MultipleColumnsCard";

describe("MultipleColumnsCard", () => {
  it("should work", () => {
    const wrapper = shallow(
      <MultipleColumnsCard gridColumns={["400px", "200px"]} cardBorder={true} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe("transformCss processor", () => {
    it("should transform gridColumns props to grid css", () => {
      const gridColumns = ["400px", 1, 4, "600px"];

      const result = transformCss(gridColumns);

      expect(result).toEqual("400px 1fr 4fr 600px ");
    });

    it("should return origin value if the gridColumns value is string", () => {
      const result = transformCss("40px");
      expect(result).toEqual("40px ");

      const result2 = transformCss("40px 1fr 500px 3fr");
      expect(result2).toEqual("40px 1fr 500px 3fr ");
    });

    it("should return default value if the gridColumns props is empty", () => {
      const result = transformCss(undefined);
      expect(result).toEqual("1fr");
    });
  });
});
