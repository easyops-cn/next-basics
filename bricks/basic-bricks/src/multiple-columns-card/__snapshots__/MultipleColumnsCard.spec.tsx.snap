// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleColumnsCard should work 1`] = `
<Card
  bordered={true}
>
  <div
    className="multiple-columns-container"
    style={
      Object {
        "gridTemplateColumns": "400px 200px ",
      }
    }
  >
    <slot
      id="contentSlot"
      name="content"
    />
  </div>
</Card>
`;
